docs: add the protocol properties document (#1130)
sig-net/mpc: doc/protocol_properties.md

# **Protocol Properties** 

Goals:

* capture *what must hold* in the MPC node coordination layer and *why* to be able to reason about simplifications that improve robustness

Protocol properties defined below are separated into

* safety (nothing bad ever happens: no key leak, no unrequested signature)  
* liveness (every request eventually completes), and  
* efficiency (we try not to waste artifacts or rounds doing it). Some waste is acceptable as long as it stays bounded.

Not all of the below is implemented. Where a property or invariant is only partly enforced today, the gap is named where it is claimed.

## 1\. Request lifecycle and its objects

* Sign request flow  
  1. request is finalized on-chain 
  2. indexed by each node's own indexer; the request\_id is a deterministic function of the request, so every node derives the same one  
  3. a per-node task organizes rounds (attempts) until one generation instance produces a signature  

>   4. the successful generation's proposer submits it on chain  

Reviewer: We may add that "in case of a failure, the signature is published by another generation participant".

Author: ```suggestion
  4. the signature is submitted on-chain
```
Thinking a bit more about it, I think it's better not to have too much detail here. We haven't introduced proposer etc yet.

>   5. settlement is chain-specific: NEAR verifies the signature and resumes the yielded promise, emitting no event of its own; the EVM contract keeps no request state and emits an event for any submitter without checking anything  

Reviewer: I would not even mention NEAR. It is not something we released. It is only used for tests, and we intend to have event-based design everywhere for consistency and simplicity.

Author: ```suggestion
  5. the contract emits an event for the submission without checking anything  
```
better?

  A request's round tries to establish a signature instance; since the artifact tentatively picked for this round can differ round to round, rounds are usually different instances. If a round is not successful, a new proposer is chosen deterministically (rotating over the nodes). Only posit messages name the round; generation messages name only (sign\_id, presignature\_id), so two rounds that re-pick the same presignature, possible when the first aborted before committing it, share one name.

## 2\. System model

* On-chain information
  * Entities: n nodes; membership, threshold t, and epoch are fixed by the NEAR governance contract. 
    * Together with the public keys and whether the contract is `Running`, they form the **committee** a node operates under (`GovernanceInfo` in code).
    * Each node holds its own local, read-only snapshot of it, refreshed on contract update. Throughout, *committee* and *committee member* are contract-scoped, while *participant* and *participant list* always refer to one instance.

>   * Request and response (per supported chain): NEAR holds pending requests, verifies signatures for them and settles them. The others mostly emit an event for whoever submits, which is why exactly-once is the consumer's job (§1).

Reviewer: Again, I would not focus on NEAR as much.

Author: ```suggestion
  * Request and response (per supported chain): 
```

  
  (i) all nodes observe the same finalized sequence of requests and committee states, possibly at different times;  
  (ii) nodes act only on finalized state (finalization definition may differ per chain).  
* Hybrid fault model.  
  * Liveness is argued under crash-recovery: correct nodes may halt and restart (otherwise never deviating). After restart the node's durable storage is assumed to be intact.  

>   * Safety of S1–S4 must hold against f arbitrarily corrupted nodes.

Reviewer: Let's define "Safety properties" before referencing them to avoid confusion.

Author: ```suggestion
  * Safety must hold against f arbitrarily corrupted nodes.
```

  * Links are fair-lossy (may drop messages), which every phase compensates for by timeout-and-retry, so during a synchronous interval, communication between live nodes is effectively reliable and timely.
  * Channels are authenticated and encrypted. 
* Each node runs its own chain indexers and eventually observes every finalized request (assumption; liveness depends on it, safety does not).  

> * Chain state reaches a node only through its RPC provider, which sits outside the fault model above: a provider that reports a committee or a request the chain never finalized breaks S3 and S2 at that node, and nothing in this layer detects it. Verifying responses against a light client is what would remove the assumption; until then it is one, and it covers the indexers as well.  

Reviewer 2: This is correct. S2 and S3 fully reliy on a trustworthy RPC provider.

@ppca @and @isSerge How much work would it be to use at lest 2 different RPC providers per chain?

Reviewer: Note: the current security model (without a light client) is to use different RPC providers on different nodes. NR < T, where NR is a number of nodes running on R RPC provider.

Author: Thanks for the clarification, I rephrased it accordingly

> * Chain state reaches a node only through its RPC provider, which sits outside the fault model above: a provider that reports a committee or a request the chain never finalized breaks S3 and S2 at that node, and nothing in this layer detects it. Verifying responses against a light client is what would remove the assumption; until then it is one, and it covers the indexers as well.  

Author: ```suggestion
* Chain state reaches a node only through its RPC provider. Different nodes can use different RPC providers. The current security model assumes that the number of nodes using a specific RPC provider is below t.  
```

Did I interpret your statement correctly.

> * The mesh active set is a local, unreliable failure detector: each node's own guess at which members are currently reachable and up-to-date (active), a subset of the committee. It may be wrong, and no two nodes ever need the same guess; §6 D2 governs what may be derived from it. A reachable peer is additionally kept *out* of the active set while an initial or post-reconnect state sync runs (a transient Syncing state), so "active" is strictly narrower than "reachable".

Reviewer: This is all true, but it may need simplification, since we are not using `active` when selecting protocol participants anymore.

Author: We still use it to determine a "good" presignature to propose.

## 3\. Safety properties

> * **S1. One-shot artifacts are consumed at most once:** No presignature is used in signature shares for more than one sign request, and no triple pair for more than one presignature.

Reviewer: I would replace "sign request" with "signature generation". We want to ensure node providers do not see 2+ generations as well.

Author: ```suggestion
* **S1. One-shot artifacts are consumed at most once:** No presignature is used in signature shares for more than one signature generation, and no triple pair for more than one presignature.
```
true, thanks for catching this!

* **S3. Membership and epochs change only on-chain:** No honest node adopts a committee other than from finalized contract state, and no node processes requests under a non-Running state. 
* **S4. Instance-local agreement:** Any two honest participants that reach the generating phase of the same instance use the same participant list or they abort.

Violating any of these is unacceptable in any execution within §2's fault bound. See Appendix for more details, including enforcing mechanisms.

## 4\. Liveness

Progress is guaranteed during a sufficiently long synchronous interval, never at a fixed time. The precise bounds and the enforcing mechanisms are in the Appendix.

* **L1. Signature progress:** during a long-enough δ-synchronous interval with ≥ t correct committee members online, a usable presignature (owner and holders online), and ≥ t of that presignature's holders having indexed the request, the request produces a signature within bounded time (O(f \+ p)·∆timeout \+ one generation round, where p counts the members that are offline or own no usable presignature). That such a presignature keeps existing is L2.  

> * **L2. Artifact supply:** during a long-enough synchronous interval with ≥ t members (itself included) in the node's active set, a node below its artifact floor and under the network-wide cap eventually completes a generation, given its inputs (none for a triple pair, one owned triple pair for a presignature).   

Reviewer: Do we still want to use `active` in our descriptions?

Author: The current implementation doesn't guarantee it if there are fewer in the active set.
We need some description of the status of the nodes in the subnet. We can use a different definition, but I don't think it helps us here.

Mechanisms striving towards these targets are deterministic proposer rotation, the posit round, and a backoff that pauses proposing when so many peers reject with "already generating" that no threshold set is left to work with.

Duplicate instances, duplicate on-chain responses, and wasted rounds are correct but wasteful. On NEAR the contract settles a request on the first valid response and later ones simply fail; on chains whose contract keeps no request state, every response emits another event and the downstream consumer is the one that has to deduplicate (§1). Consequently, any mechanism that serves only E-properties may be lossy, heuristic, or deleted; it must be judged on cost, not correctness.

## 6\. Design Invariants

Incomplete list of general, implementation-independent rules (the *how*) that keep the properties above (the *what*) true.
 
* D1. The chain is the sole arbiter. Request settlement and membership live on-chain. The off-chain layer must remain safe under arbitrary duplication, reordering, loss, and retry of its own actions. *Serves S2, S3 (safety-critical agreement is delegated to the chain) and is necessary for D4.*  

> * D2. Anything two nodes must compute identically (shared input, e.g. "who proposes in round r") may depend only on contract state, request data (e.g. the entropy, which is chain-fixed per request, the transaction hash on Ethereum), and the round number, never on the local failure detector or local clocks.  Rationale: local views differ by design; deriving shared roles from them silently reintroduces the agreement problem this layer avoids. Two caveats: shared is not the same as unbiasable, since a requester who controls the transaction controls the entropy and with it the first proposer; and "not local clocks" constrains the mapping from shared inputs to roles, not the round number itself, which each node advances on its own timeouts (nodes disagreeing about the current round is expected and is repaired by buffering future-round messages and by stale-round rejects). *Serves L1.*  

Reviewer: The requester does not control or provide the entropy. It originates from the chain.
But the current implementation of entropy selection is chain-specific and has questionable logic. Related: https://github.com/sig-net/mpc/pull/724

Author: that's not entirely correct, the transaction hash is can be influenced with nonce and calldata.. to do it better we could rely on the chain's native per-block randomness (currently droppped by indexer).

*Enforcement*: An instance is identified by (sign\_id, presignature\_id) for signing, and analogously by artifact id for triple/presignature generation. The instance's proposer alone determines the participant list and sends it in Start to each accepter; anything inconsistent is rejected and the instance aborts (§D4). Only the presignature's owner can consume it (§D3), but that check runs at take time, so a non-owner can drive a posit as far as Start; the instance then dies when every holder's take fails. No agreement *across* instances is claimed: two rounds of the same request may overlap, and safety does not depend on that never happening.

This holds even against a Byzantine proposer: A malicious proposer can send divergent lists to different honest accepters, but each honest node uses the list it was told and different nodes were told different lists, so they never reconstruct a valid signature: the combine fails and the instance aborts. The worst outcome is an abort, never a bad signature. 

### L1. Signature progress.

> *Property.* During a long-enough δ-synchronous interval with ≥ t correct committee members online, a usable presignature (owner and holders online), and ≥ t of that presignature's holders having indexed the request, the request produces a signature within bounded time (O(f \+ p)·∆timeout \+ one generation round, where p counts the members that are offline or own no usable presignature). That such a presignature keeps existing is L2.

Reviewer: I would ask AI to look for duplications in this doc; a couple of things are repeated at least 2 times.

Author: Each property is stated in the main part and then repeated in the appendix. I will go over the doc to avoid additional duplication.

∆timeout is not a constant: the round budget follows the shared schedule `round_timeout(r)` (a generous round 0, then geometric growth from a short floor to a hard ceiling), so rotating past the f \+ p proposers that cannot deliver costs the sum of `round_timeout` over the failed rounds. That sum is dominated by the round-0 budget while that count is small (20 s, then 2 s, 2.3 s, 2.6 s, ...): the ceiling is reached only after tens of rounds, so (f \+ p) times the ceiling is a bound, not an estimate.

*Rationale.* This is the property a user actually observes: a request that never reaches a signature is indistinguishable from the service being down.

*Enforcement*: 
The proposer starts as soon as every invitee has answered (Accept or Reject) and at least t have accepted, so with everyone honest and up-and-running the signature is produced at network speed in one generation round, independent of ∆timeout. Silent invitees do not hold the round hostage: 500 ms after it sends Propose the proposer goes ahead with whoever accepted, as long as that is still t. An alternative proposer gets a chance after ∆timeout.

> Proposer election is a pure function of shared inputs (`proposer_per_round` over round, membership, entropy) and the deadline `round_timeout(r)` depends only on the round `r`. Therefore peers in the same round agree on the proposer and the deadline (D2). A round advance is only ever triggered locally (deadline, enough rejects, abort), but the round it advances to is the highest a peer has shown us, so a node that falls behind catches up in one bump by learning the rejector's current round rather than climbing one round per attempt. 

Reviewer: I believe it is better to have a single description of each layer. Looks like posits will have their own doc: https://github.com/sig-net/mpc/pull/1152
Would it be a good idea to move some of the posit descriptions there and then link that doc here?

Author: Yeah, that's a good idea. I'll do that in a follow-up PR.

A node under its own floor still generates nothing while the network-wide potential is at the cap. That is intended, the pool being full, but it means the floor alone does not entail progress.

*Rationale.* L1 assumes a usable presignature is there to consume, so without supply that precondition eventually fails and signing stops while the network is otherwise perfectly healthy.

*Enforcement*: 
A node re-proposes every 100 ms while it is below its floor (`min_triples` / `min_presignatures`) and the network-wide potential is under the cap (`max_triples` / `max_presignatures`); concurrency counters free themselves as generation tasks finish and as proposals expire.  
Generation is skipped entirely while \< t nodes are in the active set, so the failure detector gates supply though never correctness.

### L3. Settlement

*Property.* Once a signature is produced with a correct, online owner, it is eventually accepted on-chain (on NEAR this must happen before the request's yield deadline).

*Rationale.* A signature that never reaches the chain is worth no more to the user than no signature at all, while the presignature spent producing it is gone either way.

> For NEAR-originated requests this requires settling before the yielded promise's bounded lifetime expires (a hard deadline of 200 blocks \~= 200s)  

Reviewer: I'm not highlighting every NEAR occurrence, but I think we need to remove all the mentions, since they are not important and only create mental load.

Author: I like that :) Done

L1 and L3 draw on that one deadline, and the rounds spend it first: round 0 costs 20 s and later rounds start at 2 s and grow by 1.15, so about 19 further rounds fit inside 200 s. A NEAR request that needs more rounds than that cannot settle at all, however the settlement path is built, which is the constraint any republish or failover schedule has to fit into.

*Enforcement*: The owner republishes on a timer, retrying indefinitely with backoff capped at 60 s.

> Currently not guaranteed, for two reasons. Only the successful instance's proposer publishes, and no other participant takes over, even though every participant holds the finished signature. And the retry itself lives in the publishing process, so what survives a restart is only what reached a confirmed checkpoint: on the checkpointed chains that includes the pending publish and its signature, while NEAR and Bitcoin produce no checkpoints at all and keep it in memory alone. On NEAR the 200 s yield deadline can also expire while the publisher is still backing off.

Reviewer 2: Hm, I'm not sure this model for NEAR works well with the no-timeout guarantee we want in bidirectional requests. I guess if we ever implement that for NEAR as the source chain, that will need to work without the yield/resume mechanism.

Author: it doesnt. Afaics, this is today's status.

Reviewer 2: Yeah it is today's status but we also don't really care about signature requests on NEAR. Something to keep in mind if traffic on NEAR becomes important later.

Reviewer: If we ever have traffic on NEAR, we can switch to an event-based model.

Author: I removed all mentions of NEAR

# **Protocol Properties** 

Draft, last updated 2026, August 12, 15:31 UTC

> 

Author: ```suggestion
```
